File: src/scaffold/infra.ts
import { existsSync } from 'node:fs'
import path from 'node:path'
import type { ConflictStrategy, FileRecord, InfraId } from '../types.js'
import { applyStrategy } from '../utils/conflict-strategy.js'
import { copyFile, ensureDir, fileExists, fileHash, readFile, writeFile } from '../utils/files.js'

export interface ScaffoldInfraOptions {
  targetDir: string
  libraryDir: string
  infra: InfraId[]
  projectName: string
  fileRecords: FileRecord[]
  strategy: ConflictStrategy
  perFileOverrides: Map<string, ConflictStrategy>
}

/**
 * Scaffolds infrastructure files into the target directory.
 *
 * Behavior:
 * - `pre-commit`: if `.git` exists at targetDir, create `.git/hooks/` and copy `library/infra/pre-commit.hook` → `.git/hooks/pre-commit`
 * - `compliance`: copy `library/infra/compliance.md` → `specs/compliance.md`
 * - `KNOWLEDGE_MAP`: read `library/infra/KNOWLEDGE_MAP.template.md`, replace `[YOUR_PROJECT_NAME]` with projectName, write to `targetDir/KNOWLEDGE_MAP.md`
 * - `codeowners`: copy `library/infra/CODEOWNERS.template` → `targetDir/CODEOWNERS`
 * - Records each written file in `fileRecords`
 * - If `infra` is empty, does nothing
 */
export async function scaffoldInfra(opts: ScaffoldInfraOptions): Promise<void> {
  const { targetDir, libraryDir, infra, projectName, fileRecords, strategy, perFileOverrides } = opts

  if (infra.length === 0) {
    return
  }

  // Process pre-commit hook
  if (infra.includes('pre-commit')) {
    if (existsSync(path.join(targetDir, '.git'))) {
      const hookDir = path.join(targetDir, '.git', 'hooks')
      ensureDir(hookDir)

      const src = path.join(libraryDir, 'infra', 'pre-commit.hook')
      const dest = path.join(hookDir, 'pre-commit')

      if (fileExists(src)) {
        await copyLibraryFile(src, dest, fileRecords, targetDir, strategy, perFileOverrides)
      }
    }
  }

  // Process compliance.md
  if (infra.includes('compliance')) {
    const specsDir = path.join(targetDir, 'specs')
    ensureDir(specsDir)

    const src = path.join(libraryDir, 'infra', 'compliance.md')
    const dest = path.join(specsDir, 'compliance.md')

    if (fileExists(src)) {
      await copyLibraryFile(src, dest, fileRecords, targetDir, strategy, perFileOverrides)
    }
  }

  // Process KNOWLEDGE_MAP
  if (infra.includes('KNOWLEDGE_MAP')) {
    const src = path.join(libraryDir, 'infra', 'KNOWLEDGE_MAP.template.md')
    const dest = path.join(targetDir, 'KNOWLEDGE_MAP.md')

    if (fileExists(src)) {
      let content = readFile(src)
      content = content.replace(/\[YOUR_PROJECT_NAME\]/g, projectName)
      await writeRootFile(dest, content, fileRecords, targetDir, 'infra/KNOWLEDGE_MAP.template.md', strategy, perFileOverrides)
    }
  }

  // Process CODEOWNERS
  if (infra.includes('codeowners')) {
    const src = path.join(libraryDir, 'infra', 'CODEOWNERS.template')
    const dest = path.join(targetDir, 'CODEOWNERS')

    if (fileExists(src)) {
      await copyLibraryFile(src, dest, fileRecords, targetDir, strategy, perFileOverrides)
    }
  }
}

/**
 * Helper: Copy a file from library with conflict resolution and recording.
 */
async function copyLibraryFile(
  src: string,
  dest: string,
  records: FileRecord[],
  targetDir: string,
  strategy: ConflictStrategy,
  perFileOverrides: Map<string, ConflictStrategy>
): Promise<void> {
  if (!fileExists(src)) return

  const relPath = path.relative(targetDir, dest)
  const action = applyStrategy(dest, strategy, perFileOverrides, targetDir)

  if (action === 'skip') {
    console.warn(`⚠️  Skipping existing file: ${relPath}`)
    return
  }

  copyFile(src, dest)
  records.push({
    path: relPath,
    hash: fileHash(dest),
    source: path.relative(targetDir, src),
  })
}

/**
 * Helper: Write content to file with conflict resolution and recording.
 */
async function writeRootFile(
  dest: string,
  content: string,
  records: FileRecord[],
  targetDir: string,
  source: string,
  strategy: ConflictStrategy,
  perFileOverrides: Map<string, ConflictStrategy>
): Promise<void> {
  const relPath = path.relative(targetDir, dest)
  const action = applyStrategy(dest, strategy, perFileOverrides, targetDir)

  if (action === 'skip') {
    console.warn(`⚠️  Skipping existing file: ${relPath}`)
    return
  }

  writeFile(dest, content)
  records.push({
    path: relPath,
    hash: fileHash(dest),
    source,
  })
}

File: internal/scaffold/infra.go
package scaffold

import (
	"log"
	"os"
	"path/filepath"
	"strings"

	"github.com/ricardoborges-teachable/ai-setup/internal/conflict"
	"github.com/ricardoborges-teachable/ai-setup/internal/files"
	"github.com/ricardoborges-teachable/ai-setup/internal/types"
)

// ScaffoldInfra installs infrastructure files (pre-commit hook, compliance,
// CODEOWNERS, KNOWLEDGE_MAP). Ported from src/scaffold/infra.ts.
func ScaffoldInfra(targetDir, libraryDir, projectName string, infra []types.InfraId, fileRecords *[]types.TrackedFile, strategy types.ConflictStrategy, perFileOverrides map[string]types.ConflictStrategy) error {
	if len(infra) == 0 {
		return nil
	}

	// Process pre-commit hook.
	for _, id := range infra {
		if id == types.InfraIdPreCommit {
			if err := scaffoldPreCommit(targetDir, libraryDir, fileRecords, strategy, perFileOverrides); err != nil {
				return err
			}
		}
	}

	// Process compliance.md.
	for _, id := range infra {
		if id == types.InfraIdCompliance {
			if err := scaffoldCompliance(targetDir, libraryDir, fileRecords, strategy, perFileOverrides); err != nil {
				return err
			}
		}
	}

	// Process KNOWLEDGE_MAP.
	for _, id := range infra {
		if id == types.InfraIdKnowledgeMap {
			if err := scaffoldKnowledgeMap(targetDir, libraryDir, projectName, fileRecords, strategy, perFileOverrides); err != nil {
				return err
			}
		}
	}

	// Process CODEOWNERS.
	for _, id := range infra {
		if id == types.InfraIdCodeowners {
			if err := scaffoldCodeowners(targetDir, libraryDir, fileRecords, strategy, perFileOverrides); err != nil {
				return err
			}
		}
	}

	return nil
}

func scaffoldPreCommit(targetDir, libraryDir string, fileRecords *[]types.TrackedFile, strategy types.ConflictStrategy, perFileOverrides map[string]types.ConflictStrategy) error {
	gitDir := filepath.Join(targetDir, ".git")
	if _, err := os.Stat(gitDir); os.IsNotExist(err) {
		return nil
	}

	hookDir := filepath.Join(gitDir, "hooks")
	if err := files.EnsureDir(hookDir); err != nil {
		return err
	}

	src := filepath.Join(libraryDir, "infra", "pre-commit.hook")
	dest := filepath.Join(hookDir, "pre-commit")

	return copyInfraFile(src, dest, targetDir, fileRecords, strategy, perFileOverrides)
}

func scaffoldCompliance(targetDir, libraryDir string, fileRecords *[]types.TrackedFile, strategy types.ConflictStrategy, perFileOverrides map[string]types.ConflictStrategy) error {
	specsDir := filepath.Join(targetDir, "specs")
	if err := files.EnsureDir(specsDir); err != nil {
		return err
	}

	src := filepath.Join(libraryDir, "infra", "compliance.md")
	dest := filepath.Join(specsDir, "compliance.md")

	return copyInfraFile(src, dest, targetDir, fileRecords, strategy, perFileOverrides)
}

func scaffoldKnowledgeMap(targetDir, libraryDir, projectName string, fileRecords *[]types.TrackedFile, strategy types.ConflictStrategy, perFileOverrides map[string]types.ConflictStrategy) error {
	src := filepath.Join(libraryDir, "infra", "KNOWLEDGE_MAP.template.md")
	if !files.FileExists(src) {
		return nil
	}

	data, err := files.ReadFile(src)
	if err != nil {
		return err
	}

	content := strings.ReplaceAll(string(data), "[YOUR_PROJECT_NAME]", projectName)
	dest := filepath.Join(targetDir, "KNOWLEDGE_MAP.md")
	relPath, err := filepath.Rel(targetDir, dest)
	if err != nil {
		relPath = dest
	}

	action, err := conflict.ApplyStrategy(dest, strategy, perFileOverrides, targetDir)
	if err != nil {
		return err
	}
	if action == "skip" {
		log.Printf("Skipping existing file: %s", relPath)
		return nil
	}

	if err := files.WriteFile(dest, []byte(content), 0o644); err != nil {
		return err
	}

	hash, _ := files.FileHash(dest)
	*fileRecords = append(*fileRecords, types.TrackedFile{
		Path:   relPath,
		Hash:   hash,
		Source: "infra/KNOWLEDGE_MAP.template.md",
		Owner:  types.FileOwnerLibrary,
	})
	return nil
}

func scaffoldCodeowners(targetDir, libraryDir string, fileRecords *[]types.TrackedFile, strategy types.ConflictStrategy, perFileOverrides map[string]types.ConflictStrategy) error {
	src := filepath.Join(libraryDir, "infra", "CODEOWNERS.template")
	dest := filepath.Join(targetDir, "CODEOWNERS")

	return copyInfraFile(src, dest, targetDir, fileRecords, strategy, perFileOverrides)
}

// copyInfraFile copies an infrastructure file with conflict resolution and tracking.
func copyInfraFile(src, dest, targetDir string, fileRecords *[]types.TrackedFile, strategy types.ConflictStrategy, perFileOverrides map[string]types.ConflictStrategy) error {
	if !files.FileExists(src) {
		return nil
	}

	relPath, err := filepath.Rel(targetDir, dest)
	if err != nil {
		relPath = dest
	}

	action, err := conflict.ApplyStrategy(dest, strategy, perFileOverrides, targetDir)
	if err != nil {
		return err
	}
	if action == "skip" {
		log.Printf("Skipping existing file: %s", relPath)
		return nil
	}

	if err := files.CopyFile(src, dest); err != nil {
		return err
	}

	hash, _ := files.FileHash(dest)
	source, _ := filepath.Rel(targetDir, src)
	*fileRecords = append(*fileRecords, types.TrackedFile{
		Path:   relPath,
		Hash:   hash,
		Source: source,
		Owner:  types.FileOwnerLibrary,
	})
	return nil
}
